Sam irv backend tests

Sam-Irv opened the pull request:
Added a bunch of backend tests, with more to come, as well as tweaked a couple of bits of code for efficiency and possibly preventing future bugs.

Sam-Irv committed:
Update database_functions.py

Sam-Irv committed:
Update card_logic.py

Sam-Irv committed:
Update database_functions.py

Sam-Irv committed:
Back-end tests + minor bugfix

Added a few Back-end tests (focused on process_card), and tweaked a couple things which caused issues these tests raised.

mmdonaldson commented:
@jacobstringer can you please review this one

jacobstringer commented on tests/test_controllers.py:
```diff
@@ -0,0 +1,174 @@
+        
+        round3 = Round(cardId=stockade.id, playerId=player.id, round=3)
+        db_committing_function(round3)
+        self.assertTrue(process_card(stockade,player,True,False))
```
It would also be worth asserting that the player's money increases by 3 when they discard, and in the above test, that their other resources also change with the played cards.

jacobstringer reviewed (approved):
Looks good Sam.

One thing I noticed is that you declared cardhist objects for each test. I don't think any of the tests require cardhist objects in order to run, so I think you could remove them without any issue there to make it look simpler.

Sam-Irv committed:
Removed clutter, strengthened tests

Removed unnecessary 'defaults' section, removed unneeded Cardhist() objects, and added checks to make sure process card added the benefits to the player when necessary.

Sam-Irv committed:
Added missing code from last commit

Sam-Irv commented:
Noted. Strengthened the tests, removed clutter.
